### PR TITLE
update System.IdentityModel.Tokens.Jwt

### PR DESCRIPTION
**Pull request by omsmith**

### CODE REVIEW

File: src/D2L.Security.OAuth2/Keys/Default/D2LSecurityToken.cs [hunk 1]
```diff
@@ -27,96 +23,80 @@ internal sealed partial class D2LSecurityToken : SecurityToken {
...
 		}
 
-		public string KeyId { get { return m_id; } }
-
-		public override DateTime ValidFrom {
-			get { return m_validFrom.UtcDateTime; }
-		}
+		public string Id { get; }
+		public DateTimeOffset ValidFrom { get; }
+		public DateTimeOffset ValidTo { get; }
```
Comment (omsmith) [inline src/D2L.Security.OAuth2/Keys/Default/D2LSecurityToken.cs:42]: - [ ] I don't believe anything checks this anymore. Need to assess checking dates before using the key.
  - [ ] Compare to old library

File: test/D2L.Security.OAuth2.UnitTests/Keys/Default/TokenSignerTests.cs [hunk 2]
```diff
@@ -11,8 +11,8 @@ namespace D2L.Security.OAuth2.Keys.Default {
...
+        private const string SignedComplexToken = "eyJhbGciOiJSUzI1NiIsImtpZCI6ImZhN2MwN2E4LTQyYzgtNGM1Ny05YWYyLWNjZTEwYzI3MTAzMyIsInR5cCI6IkpXVCJ9.eyJuYmYiOjE1NDYzMDA4MDAsImV4cCI6MTU0NjMwNDQzMCwiaXNzIjoiaXNzdWVyIiwiYXVkIjoiYXVkaWVuY2UiLCJodHRwczovL3NlcnZpY2UuY29tL2NsYWltL3ZlcnNpb24iOiIxLjAuMCIsImh0dHBzOi8vc2VydmljZS5jb20vY2xhaW0vcm9sZXMiOlsiaHR0cHM6Ly9zZXJ2aWNlLmNvbS9yb2xlcyNhZG1pbmlzdHJhdG9yIiwiaHR0cHM6Ly9zZXJ2aWNlLmNvbS9yb2xlcyN1c2VyIl0sImh0dHBzOi8vc2VydmljZS5jb20vY2xhaW0vY29udGV4dCI6eyJpZCI6ImMxZDg4N2YwLWExYTMtNGJjYS1hZTI1LWMzNzVlZGNjMTMxYSIsInR5cGUiOlsiaHR0cHM6Ly9zZXJ2aWNlLmNvbS90eXBlcyN0eXBlIl19fQ.eMvY5Srt0I-tQ-4gWi5EXmZ897IVymLSywUd6DI6QYcD4zCVoNMus7VbwcJm8P3Sb6-JUHS7ZMBuaA3E0xG4xmPtrnCUkEETfKgjQmmHGHwx0ZoBnsjYhTRUk_imf5DMPWIUS8S01QYz1pbFhxsKQGX2h-dTpkPa_PH2vCd-5TlKKGrOCkV4c60RpkQuj1UVcxpHpCecPGT-ulIA7loK2mCSov5lk_hvnmvFTH-F0eOf-CI9ebDvfWCSnKGlsBp6TAo9PIDDtH4GiwLrW6ZsSGMXGE5uWLFDGtZq-OETV6EMEOA-N4MJJ_34tJ0SPVSJ_8ApyP8_HBzGb9ED8PQBOQ";
```
Comment (omsmith) [inline test/D2L.Security.OAuth2.UnitTests/Keys/Default/TokenSignerTests.cs:15]: Ordering of the JSON keys in the header in payload changed. Checked they were equivalent values though.

File: test/D2L.Security.OAuth2.UnitTests/Validation/AccessTokenValidatorTests.cs [hunk 3]
```diff
@@ -36,6 +38,7 @@ public class AccessTokenValidatorTests {
 
 		[Test]
 		public async Task ExpiredJwt() {
+			IdentityModelEventSource.ShowPII = true;
```
Comment (omsmith) [inline test/D2L.Security.OAuth2.UnitTests/Validation/AccessTokenValidatorTests.cs:41]: - [ ] Exceptions aren't very useful without this. Set for tests generally.

**Comment by omsmith** [comment]
de0944fdf05027a849e6dc2ec47684047c6c43ef (this branch)

``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.18362.1082 (1903/May2019Update/19H1)
Intel Core i7-9700K CPU 3.60GHz (Coffee Lake), 1 CPU, 8 logical and 8 physical cores
  [Host]     : .NET Framework 4.8 (4.8.4220.0), X64 RyuJIT
  DefaultJob : .NET Framework 4.8 (4.8.4220.0), X64 RyuJIT


```
| Method |        Mean |    Error |   StdDev | Rank |
|------- |------------:|---------:|---------:|-----:|
|  ES256 |   277.74 μs | 1.396 μs | 1.238 μs |    2 |
|  ES384 |   602.93 μs | 2.869 μs | 2.684 μs |    3 |
|  ES512 | 1,300.85 μs | 1.639 μs | 1.280 μs |    4 |
|  RS256 |    87.30 μs | 0.757 μs | 0.671 μs |    1 |

**Comment by omsmith** [comment]
f382c8dd08e4350ad16fa35e5dff69397e81923a (master)

``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.18362.1082 (1903/May2019Update/19H1)
Intel Core i7-9700K CPU 3.60GHz (Coffee Lake), 1 CPU, 8 logical and 8 physical cores
  [Host]     : .NET Framework 4.8 (4.8.4220.0), X64 RyuJIT
  DefaultJob : .NET Framework 4.8 (4.8.4220.0), X64 RyuJIT


```
| Method |       Mean |   Error |  StdDev | Rank |
|------- |-----------:|--------:|--------:|-----:|
|  ES256 |   316.4 μs | 2.46 μs | 2.06 μs |    2 |
|  ES384 |   647.4 μs | 2.32 μs | 2.17 μs |    3 |
|  ES512 | 1,343.3 μs | 3.15 μs | 2.63 μs |    4 |
|  RS256 |   134.8 μs | 0.49 μs | 0.46 μs |    1 |

File: src/D2L.Security.OAuth2/Keys/Default/D2LSecurityToken.cs (hunk 1 above)
Comment (omsmith) [reply src/D2L.Security.OAuth2/Keys/Default/D2LSecurityToken.cs:42]: Initial assessment says the old library didn't look at these either. So we'll sign/verify with keys that are out of range.

I believe our key cache stuff negates this, but, something to assess / throw an error for down the line?

File: src/D2L.Security.OAuth2/Keys/Default/D2LSecurityToken.cs [hunk 4]
```diff
@@ -1,15 +1,11 @@
...
+using Microsoft.IdentityModel.Tokens;
 
 namespace D2L.Security.OAuth2.Keys.Default {
 
-	internal sealed partial class D2LSecurityToken : SecurityToken {
-
-		private string m_id;
-		private readonly DateTimeOffset m_validFrom;
-		private readonly DateTimeOffset m_validTo;
+	internal sealed partial class D2LSecurityToken : SecurityKey {
```
Comment (omsmith) [inline src/D2L.Security.OAuth2/Keys/Default/D2LSecurityToken.cs:8]: This should get renamed `D2LSecurityKey` for less confusion, but trying not to bloat the diff more.

File: src/D2L.Security.OAuth2/Keys/Default/D2LCryptoProviderFactory.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
+﻿using Microsoft.IdentityModel.Tokens;
+
+namespace D2L.Security.OAuth2.Keys.Default {
+	internal sealed class D2LCryptoProviderFactory : CryptoProviderFactory {
```
Comment (omsmith) [inline src/D2L.Security.OAuth2/Keys/Default/D2LCryptoProviderFactory.cs:4]: `SecurityKey` as a `CryptoProviderFactory` property, but it's not overridable, so this needs to be provided to `SigningCredentials` / `TokenValidationParameters`

File: src/D2L.Security.OAuth2/Keys/Default/EcDsaJsonWebKey.ECCPublicKeyBlobFormatter.cs [hunk 6]
```diff
@@ -1,5 +1,6 @@
 ﻿using System;
```
Comment (omsmith) [inline src/D2L.Security.OAuth2/Keys/Default/EcDsaJsonWebKey.ECCPublicKeyBlobFormatter.cs:1]: I _believe_ they've added stuff that we could migrate to and make this unnecessary, will assess after.

File: src/D2L.Security.OAuth2/Keys/Default/D2LSecurityToken.cs (hunk 1 above)
Comment (j3parker) [reply src/D2L.Security.OAuth2/Keys/Default/D2LSecurityToken.cs:42]: Ouch, yeah :(

**Review by j3parker — APPROVED** [review]

**Pull request merged**